File: eco-core/core-plugin/src/main/java/com/willfp/ecoskills/EcoSkillsPlugin.java
package com.willfp.ecoskills;

import com.willfp.eco.core.AbstractPacketAdapter;
import com.willfp.eco.core.Prerequisite;
import com.willfp.eco.core.command.impl.PluginCommand;
import com.willfp.eco.core.integrations.IntegrationLoader;
import com.willfp.ecoskills.actionbar.ActionBarClearOnGamemode;
import com.willfp.ecoskills.actionbar.ActionBarCompatChatMessage;
import com.willfp.ecoskills.actionbar.ActionBarCompatSetActionBar;
import com.willfp.ecoskills.actionbar.ActionBarUtils;
import com.willfp.ecoskills.actionbar.HealthScaleListener;
import com.willfp.ecoskills.commands.CommandEcoSkills;
import com.willfp.ecoskills.commands.CommandSkills;
import com.willfp.ecoskills.config.EffectsYml;
import com.willfp.ecoskills.data.DataListener;
import com.willfp.ecoskills.data.LeaderboardHandler;
import com.willfp.ecoskills.effects.CustomEffectUtils;
import com.willfp.ecoskills.effects.CustomEffects;
import com.willfp.ecoskills.effects.Effect;
import com.willfp.ecoskills.effects.Effects;
import com.willfp.ecoskills.integrations.EcoEnchantsEnchantingLeveller;
import com.willfp.ecoskills.integrations.enchantgui.EnchantGuiHandler;
import com.willfp.ecoskills.placeholders.EcoSkillsTopPlaceholder;
import com.willfp.ecoskills.skills.CustomSkills;
import com.willfp.ecoskills.skills.Skill;
import com.willfp.ecoskills.skills.SkillDisplayListener;
import com.willfp.ecoskills.skills.SkillLevellingListener;
import com.willfp.ecoskills.skills.Skills;
import com.willfp.ecoskills.stats.CustomStatUtils;
import com.willfp.ecoskills.stats.CustomStats;
import com.willfp.ecoskills.stats.DamageIndicatorListener;
import com.willfp.ecoskills.stats.Stat;
import com.willfp.ecoskills.stats.Stats;
import com.willfp.ecoskills.stats.modifier.StatModifierListener;
import com.willfp.libreforge.HolderProvider;
import com.willfp.libreforge.HolderProviderKt;
import com.willfp.libreforge.ProvidedHolder;
import com.willfp.libreforge.SimpleProvidedHolder;
import com.willfp.libreforge.loader.LibreforgePlugin;
import com.willfp.libreforge.loader.configs.ConfigCategory;
import org.bukkit.event.Listener;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.List;

public class EcoSkillsPlugin extends LibreforgePlugin {
    /**
     * Instance of EcoSkills.
     */
    private static EcoSkillsPlugin instance;

    /**
     * effects.yml.
     */
    private final EffectsYml effectsYml;

    /**
     * Internal constructor called by bukkit on plugin load.
     */
    public EcoSkillsPlugin() {
        instance = this;
        effectsYml = new EffectsYml(this);

        HolderProviderKt.registerHolderProvider((HolderProvider) player -> CustomEffectUtils.getCustomEffects(player).stream().map(
                it -> (ProvidedHolder) new SimpleProvidedHolder(it)
        ).toList());

        HolderProviderKt.registerHolderProvider((HolderProvider) player -> CustomStatUtils.getCustomStats(player).stream().map(
                it -> (ProvidedHolder) new SimpleProvidedHolder(it)
        ).toList());
    }

    @NotNull
    @Override
    public List<ConfigCategory> loadConfigCategories() {
        return List.of(
                new CustomStats(),
                new CustomEffects(),
                new CustomSkills()
        );
    }

    @Override
    public void handleEnable() {
        EcoSkillsTopPlaceholder.register(this);

        Skills.update(this);
    }

    @Override
    public void handleReload() {
        for (Effect effect : Effects.values()) {
            this.getEventManager().unregisterListener(effect);
            this.getEventManager().registerListener(effect);
        }
        for (Stat stat : Stats.values()) {
            this.getEventManager().unregisterListener(stat);
            this.getEventManager().registerListener(stat);
        }
        for (Skill skill : Skills.values()) {
            this.getEventManager().unregisterListener(skill);
            this.getEventManager().registerListener(skill);
        }

        if (this.getConfigYml().getBool("persistent-action-bar.enabled")) {
            ActionBarUtils.startRunnable();
        }

        SkillDisplayListener.tickBossBars(this);
        this.getScheduler().runAsyncTimer(new LeaderboardHandler.Runnable(), 50, 2400);
    }

    /**
     * Get effects.yml.
     *
     * @return effects.yml.
     */
    public EffectsYml getEffectsYml() {
        return effectsYml;
    }

    /**
     * Get the instance of EcoSkills.
     *
     * @return Instance.
     */
    public static EcoSkillsPlugin getInstance() {
        return instance;
    }

    @Override
    protected List<Listener> loadListeners() {
        return Arrays.asList(
                new SkillLevellingListener(),
                new SkillDisplayListener(this),
                new StatModifierListener(),
                new DataListener(),
                new DamageIndicatorListener(this),
                new HealthScaleListener(this),
                new ActionBarClearOnGamemode()
        );
    }

    @SuppressWarnings("deprecation")
    @Override
    protected List<AbstractPacketAdapter> loadPacketAdapters() {
        if (Prerequisite.HAS_1_19.isMet()) {
            return List.of(
                    new ActionBarCompatSetActionBar(this)
            );
        }

        return Arrays.asList(
                new ActionBarCompatChatMessage(this),
                new ActionBarCompatSetActionBar(this)
        );
    }

    @Override
    protected List<PluginCommand> loadPluginCommands() {
        return Arrays.asList(
                new CommandEcoSkills(this),
                new CommandSkills(this)
        );
    }

    @NotNull
    @Override
    public List<IntegrationLoader> loadIntegrationLoaders() {
        return List.of(
                new IntegrationLoader("EcoEnchants", () -> this.getEventManager().registerListener(new EcoEnchantsEnchantingLeveller(this))),
                new IntegrationLoader("EnchantGui", () -> this.getEventManager().registerListener(new EnchantGuiHandler()))
        );
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/ecoskills/EcoSkillsPlugin.kt
package com.willfp.ecoskills

import com.willfp.eco.core.command.impl.PluginCommand
import com.willfp.eco.core.packet.PacketListener
import com.willfp.ecoskills.actionbar.ActionBarCompatibilityProxy
import com.willfp.ecoskills.actionbar.ActionBarGamemodeListener
import com.willfp.ecoskills.actionbar.ActionBarHandler
import com.willfp.ecoskills.actionbar.HealthScaleDisabler
import com.willfp.ecoskills.commands.CommandEcoSkills
import com.willfp.ecoskills.commands.CommandSkills
import com.willfp.ecoskills.commands.CommandStats
import com.willfp.ecoskills.effects.Effects
import com.willfp.ecoskills.libreforge.*
import com.willfp.ecoskills.magic.MagicHandler
import com.willfp.ecoskills.magic.MagicListener
import com.willfp.ecoskills.magic.MagicTypes
import com.willfp.ecoskills.skills.EcoSkillsSkillTopPlaceholder
import com.willfp.ecoskills.skills.EcoSkillsTopPlaceholder
import com.willfp.ecoskills.skills.SkillCritListener
import com.willfp.ecoskills.skills.Skills
import com.willfp.ecoskills.skills.display.DamageIndicatorListener
import com.willfp.ecoskills.skills.display.GainXPDisplay
import com.willfp.ecoskills.skills.display.LevelUpDisplay
import com.willfp.ecoskills.skills.display.TemporaryBossBarHandler
import com.willfp.ecoskills.skills.isInDisabledWorld
import com.willfp.ecoskills.stats.StatModifierListener
import com.willfp.ecoskills.stats.Stats
import com.willfp.libreforge.SimpleProvidedHolder
import com.willfp.libreforge.conditions.Conditions
import com.willfp.libreforge.filters.Filters
import com.willfp.libreforge.loader.LibreforgePlugin
import com.willfp.libreforge.loader.configs.ConfigCategory
import com.willfp.libreforge.registerSpecificHolderProvider
import com.willfp.libreforge.triggers.Triggers
import org.bukkit.entity.Player
import org.bukkit.event.Listener

internal lateinit var plugin: EcoSkillsPlugin
    private set

class EcoSkillsPlugin : LibreforgePlugin() {
    init {
        plugin = this
    }

    override fun loadConfigCategories(): List<ConfigCategory> {
        return listOf(
            MagicTypes,
            Stats,
            Effects,
            Skills
        )
    }

    override fun handleEnable() {
        registerSpecificHolderProvider<Player> { player ->
            if (player.isInDisabledWorld) emptyList() else
                (Effects.values() union Stats.values())
                    .mapNotNull { it.getLevelHolder(it.getActualLevel(player)) }
                    .map { SimpleProvidedHolder(it) }
        }

        com.willfp.libreforge.effects.Effects.register(EffectAddStat)
        com.willfp.libreforge.effects.Effects.register(EffectMultiplyStat)
        com.willfp.libreforge.effects.Effects.register(EffectSkillXpMultiplier)
        com.willfp.libreforge.effects.Effects.register(EffectGiveSkillXp)
        com.willfp.libreforge.effects.Effects.register(EffectMultiplyAllStats)
        com.willfp.libreforge.effects.Effects.register(EffectMakeSkillCrit)
        com.willfp.libreforge.effects.Effects.register(EffectGiveMagic)
        com.willfp.libreforge.effects.Effects.register(EffectMultiplyMagic)
        com.willfp.libreforge.effects.Effects.register(EffectGiveSkillXpNaturally)
        com.willfp.libreforge.effects.Effects.register(EffectAddStatTemporarily(this))
        com.willfp.libreforge.effects.Effects.register(EffectMultiplyStatTemporarily(this))
        com.willfp.libreforge.effects.Effects.register(EffectMagicRegenMultiplier)
        Conditions.register(ConditionStatAbove)
        Conditions.register(ConditionStatBelow)
        Conditions.register(ConditionStatEquals)
        Conditions.register(ConditionHasSkillLevel)
        Conditions.register(ConditionBelowMagic)
        Conditions.register(ConditionAboveMagic)
        Triggers.register(TriggerGainSkillXp)
        Triggers.register(TriggerLevelUpSkill)
        Triggers.register(TriggerRegenMagic)
        Filters.register(FilterSkill)
        Filters.register(FilterSkillCrit)
        Filters.register(FilterMagicType)

        EcoSkillsTopPlaceholder(this).register()
        EcoSkillsSkillTopPlaceholder(this).register()
        Skills.registerPlaceholders(this)
    }

    override fun handleReload() {
        if (this.configYml.getBool("persistent-action-bar.enabled")) {
            ActionBarHandler(this).startTicking()
        }

        TemporaryBossBarHandler(this).startTicking()
        MagicHandler(this).startTicking()
    }

    override fun loadPluginCommands(): List<PluginCommand> {
        return listOf(
            CommandEcoSkills(this),
            CommandSkills(this),
            CommandStats(this)
        )
    }

    override fun loadListeners(): List<Listener> {
        return listOf(
            LevelUpDisplay(this),
            GainXPDisplay(this),
            DamageIndicatorListener(this),
            MagicListener(this),
            HealthScaleDisabler(this),
            ActionBarGamemodeListener,
            SkillCritListener,
            StatModifierListener
        )
    }

    override fun loadPacketListeners(): List<PacketListener> {
        return listOf(
            getProxy(ActionBarCompatibilityProxy::class.java)
        )
    }
}
